src/logging.c

```diff
@@ -0,0 +1,556 @@
+                  const char *content,
+                  gsize       content_len,
+                  gsize       chunk_len,
+                  int        *msgc)
+{
+    SoupMessage **msgv;
+    SoupMessage **msg;
+    gsize         content_left;
+
+    g_return_val_if_fail (uri != NULL, NULL);
+    g_return_val_if_fail (content != NULL && content_len > 0, NULL);
+    g_return_val_if_fail (chunk_len > 0, NULL);
+    g_return_val_if_fail (msgc != NULL, NULL);
+
+    *msgc = content_len / chunk_len + (content_len % chunk_len > 0);
+    msgv = g_malloc0 (sizeof (SoupMessage *) * *msgc);
+
+    msg = msgv;
+    content_left = content_len;
+
+    while (content_left > 0) {
+        gsize   msg_len;
+        goffset start;
+        goffset end;
+
+        msg_len = content_left > chunk_len ? chunk_len : content_left;
+        start = content_len - content_left;
+        end = start + msg_len - 1;
+
+        *msg = soup_message_new_from_uri ("PUT", uri);
+        soup_message_headers_set_content_range ((*msg)->request_headers, start, end, content_len);
+        soup_message_headers_append ((*msg)->request_headers, "log-level", "2");
+        soup_message_set_request (*msg, "text/plain", SOUP_MEMORY_TEMPORARY, &content[start], msg_len);
+
+        content_left -= msg_len;
+        msg++;
+    }
+
+    return msgv;
+}
+
+static void
+rstrnt_upload_log (const RstrntTask    *task,
+                   RstrntServerAppData *app_data,
+                   SoupSession         *session,
+                   GCancellable        *cancellable,
+                   RstrntLogType        type)
+{
+    RstrntLogManager *manager;
+    g_autoptr (GError) error = NULL;
+    RstrntTaskLogData *data;
+    g_autofree char *path = NULL;
+    GMappedFile *file;
+    SoupURI *uri = NULL;
+    g_autofree SoupMessage **msgv = NULL;
+    int msgc = 0;
+    const char *contents;
+    size_t length;
+    GFile *data_file = NULL;
+    gchar *log_path = NULL;
+
+    manager = rstrnt_log_manager_get_instance ();
+    data = rstrnt_log_manager_get_task_data (manager, task, &error);
+
+    g_return_if_fail (NULL != data);
+
+    switch (type)
+    {
+        case RSTRNT_LOG_TYPE_TASK:
+        {
+            data_file = data->task_log_data->file;
+            log_path = LOG_PATH_TASK;
+        }
+        break;
+
+        case RSTRNT_LOG_TYPE_HARNESS:
+        {
+            data_file = data->harness_log_data->file;
+            log_path = LOG_PATH_HARNESS;
+        }
+        break;
+    }
+
+    path = g_file_get_path (data_file);
+    file = g_mapped_file_new (path, false, &error);
+
+    if (NULL == file)
+    {
+        g_warning ("Task log file mapping failed: %s", error->message);
+
+        g_return_if_reached ();
+    }
+
+    uri = soup_uri_new_with_base (task->task_uri, log_path);
+    contents = g_mapped_file_get_contents (file);
+    length = g_mapped_file_get_length (file);
+
+    msgv = rstrnt_chunk_log (uri, contents, length, BKR_MAX_CONTENT_LENGTH, &msgc);
```

Comment:
Once we've copied the file into memory, we do it again here.

Reply:
> Once we've copied the file into memory, we do it again here.
We are not copying here. The content for the message references the mapped file contents, notice the `SOUP_MEMORY_TEMPORARY`.
https://developer.gnome.org/libsoup/stable/SoupMessageBody.html#SoupMemoryUse

> These files can get pretty big.

How big?

Reply:
The size of the file is dependent on the user's job. We don't have control over that so I feel a more conservative method
is more appropriate especially since there has been concern about the amount of memory restraint is using.

Reply:
As commented before, we aren't reading the whole file into memory and the buffer for the message references the contents of the mapped file, not copying.